Allow passing different arguments to different postprocessors

pukkandan opened the pull request:
Stripped down to minimum required code from:
https://github.com/pukkandan/yt-dlc/commit/1b77b347d422ed70fd833a9f0327ea418ba4919c

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [ ] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Allow passing different arguments to different postprocessors like this:

    --postprocessor-args "VideoConvertor:-c:v h264_nvenc -preset slow" --postprocessor-args "Metadata:-dn"

For backward compatibility, `--postprocessor-args args` is equivalent to:
    --post-processor-args "default:<args>"

This is a stripped down version of the [full patch](https://github.com/pukkandan/yt-dlc/commit/1b77b347d422ed70fd833a9f0327ea418ba4919c) in hopes that keeping the code to the bare minimum would allow this to be merged quicker

Closes https://github.com/ytdl-org/youtube-dl/issues/27593

dstftw commented on youtube_dl/postprocessor/common.py:
```diff
@@ -62,7 +64,10 @@ def try_utime(self, path, atime, mtime, errnote='Cannot update utime of file'):
+        args = self._downloader.params.get('postprocessor_args', {})
+        if isinstance(args, list):  # for backward compatibility
+            args = {'default': args}
+        return cli_configuration_args(args, self.PP_NAME.lower(), args.get('default', []))
```
`cli_configuration_args` is supposed to be used with `params` dict, if you are working with `postprocessor_args` on your own you should not use it.

pukkandan replied:
what should I use here then?

dstftw replied:
Something like this (not tested):
```python
args = self._downloader.params.get('postprocessor_args')
if args is None:
    return default
if isinstance(args, (list, tuple)):  # for backward compatibility
    return args
assert isinstance(args, dict)
pp_args = args.get(pp_key)
if pp_args is not None:
    return pp_args
pp_args = args.get('default')  # for backward compatibility
if pp_args is not None:
    return pp_args
return default
```

pukkandan replied:
That is pretty much a copy-paste of `cli_configuration_args`. Anyway, I have changed it

dstftw replied:
Actually, it's not. You still ignore `default` parameter of `_configuration_args`.

dstftw reviewed (changes requested):
It's also could be beneficial to have common `args` for each post-processor entry in `postprocessors` list for providing arguments programmatically for post-processors known beforehand right away without bothering with `postprocessor_args`.

pukkandan commented on youtube_dl/postprocessor/ffmpeg.py:
```diff
@@ -53,6 +53,8 @@ class FFmpegPostProcessorError(PostProcessingError):
 
 class FFmpegPostProcessor(PostProcessor):
     def __init__(self, downloader=None):
+        if not hasattr(self, 'PP_NAME'):
+            self.PP_NAME = self.__class__.__name__[6:-2]  # Remove ffmpeg from the front
```
I set this as a general rule for the ffmpeg PP which can be overridden by the child class i needed. For example, `EmbedThumbnail` does not start with `Ffmpeg`, so I explicitly defined the name here. If you don;t want it this way, there are 2 options:

1. Use the full class name as PP_NAME. So `ExtractAudio` will become `FFmpegExtractAudio`
2. Define the PP_NAME in each Postprocessor individually

Let me know which of these I should do

dstftw replied:
`EmbedThumbnail` does not start with ffmpeg as it uses not only ffmpeg.
Either option is fine - this is just a human readable name after all.

pukkandan replied:
nvm, you already clarified below that the key name should be the same as that when using API. That would automatically mean I have to go with option 1

pukkandan commented on youtube_dl/postprocessor/common.py:
```diff
@@ -62,7 +64,10 @@ def try_utime(self, path, atime, mtime, errnote='Cannot update utime of file'):
+        args = self._downloader.params.get('postprocessor_args', {})
+        if isinstance(args, list):  # for backward compatibility
+            args = {'default': args}
+        return cli_configuration_args(args, self.PP_NAME.lower(), args.get('default', []))
```
I dont see any reason to have both a pp_key and a pp_name when both serve the same purpose. Or are you asking me to change the pp_name implementation to be like ie_key?

pukkandan replied:
I have implemented `pp_key()` similar to `ie_key()` and removed `PP_NAME`

dstftw replied:
Keys are used for identification, names are used for display. Here key is at least required, name is optional but may be useful for supported postprocessors page generation and for overall symmetry with info extractor API.

pukkandan replied:
Name is not necessary for this PR though. So it can be implemented separately if needed

pukkandan commented on youtube_dl/__init__.py:
```diff
@@ -307,9 +308,19 @@ def parse_retries(retries):
+    postprocessor_args = {}
+    if opts.postprocessor_args is not None:
+        for string in opts.postprocessor_args:
+            mobj = re.match(r'(?P<pp>\w+):(?P<args>.*)$', string)
```
I am not sure how to do this. Could you please make this change yourself

dstftw replied:
Will do.

pukkandan commented on youtube_dl/options.py:
```diff
@@ -795,9 +795,14 @@ def _comma_separated_values_options_callback(option, opt_str, value, parser):
-        '--postprocessor-args',
-        dest='postprocessor_args', metavar='ARGS',
-        help='Give these arguments to the postprocessor')
+        '--postprocessor-args', metavar='NAME:ARGS',
```
done

pukkandan commented on youtube_dl/options.py:
```diff
@@ -795,9 +795,14 @@ def _comma_separated_values_options_callback(option, opt_str, value, parser):
+            'Give these arguments to the postprocessors. '
+            "Specify the postprocessor name and the arguments separated by a colon ':' "
+            'to give the argument to only the specified postprocessor. Supported names are '
+            'ExtractAudio, VideoConvertor, EmbedSubtitle, Metadata, Merger, FixupStretched, FixupM4a, FixupM3u8, SubtitlesConvertor and Default'
```
done. I also realized that I forgot some PP here. Added them too

pukkandan commented on youtube_dl/postprocessor/common.py:
```diff
@@ -62,7 +64,10 @@ def try_utime(self, path, atime, mtime, errnote='Cannot update utime of file'):
+        args = self._downloader.params.get('postprocessor_args', {})
+        if isinstance(args, list):  # for backward compatibility
+            args = {'default': args}
+        return cli_configuration_args(args, self.PP_NAME.lower(), args.get('default', []))
```
In that case, I assume the PP are already designed to handle `None` as its argument. 

So I have converted `None => {`default`: None}`. Is that sufficient? Should I change it to `default: []`?

pukkandan commented:
> It's also could be beneficial to have common args for each post-processor entry in postprocessors list for providing arguments programmatically for post-processors known beforehand right away without bothering with postprocessor_args.

@dstftw I assume you are talking about API access. In that case, it would be beyond the scope of this PR

remitamine commented:
i think that instead of passing argument to all post-processors should no longer be encouraged(with the `default` prefix), instead there should be a prefix for every supported program(`ffmpeg`, `AtomicParsley`, etc...) that will pass arguments to all post-processors using the same program.

dstftw commented:
One may need to pass different args to different post-processors based on the same program.

remitamine commented:
just to clarify my requested change, instead of introducing and encouraging users to use `default` prefix, i think it would be better to have `ffmpeg`, `AtomicParsley`, etc... prefixes.
it's only about the `default` prefix, it's not related to the invidual post-processor prefixes `FFmpegExtractAudio`, `EmbedThumbnail`, etc...

dstftw commented:
Suggesting using `ffmpeg` prefix will have the same effect of argument clashes as we have now with applying args to every post-processor so I don't think it's a good idea. As for default, I agree that it should not be mentioned.

remitamine commented:
> Suggesting using ffmpeg prefix will have the same effect of argument clashes as we have now with applying args to every post-processor so I don't think it's a good idea.

this happened when there wasn't another way to specify arguments for a specific post-processor, so when both posibilities are available the user should:
- use general program prefix(ex: `ffmpeg`) when he intends to pass a specific parameter to all program invocation(ex: global options such as `--log-level`, `-y`, `-n`, ...) or when the post-processor is using multiple programs.
- use specifc post-processor when he want to pass the arguments only to a specific post-processor.

while the `default` prefix is more prone to causing problems(sending arguments to different programs will cause some of them to fail).

pukkandan commented:
@remitamine I get your point and I have an idea how to implement this. But first we need to decide what should happen if the user passes both a `ffmpeg` and an `FFmpegExtractAudio`, for example. Should both set of options be used? or only ine should be prefered.

> while the default prefix is more prone to causing problems(sending arguments to different programs will cause some of them to fail).

True, but sending the args to all PP is the current behaviour. So we must maintain an option to still do it. We could maybe discourage the user from using default by either:

1. Providing a warning when default is used
2. Leaving the default option undocumented
3. Removing the default prefix. Note that the default args can still be set by using the older `--postprocessor-args ARGS` syntax

remitamine commented:
> @remitamine I get your point and I have an idea how to implement this. But first we need to decide what should happen if the user passes both a `ffmpeg` and an `FFmpegExtractAudio`, for example. Should both set of options be used? or only ine should be prefered.

if it's going to be added, i think both arguments should be used, if a user uses both the `EmbedThumbnail` and `ffmpeg` prefix then both arguments should be passed to the ffmpeg command.
anyway, adding the program suffixes is just a suggestion.
> > while the default prefix is more prone to causing problems(sending arguments to different programs will cause some of them to fail).
> 
> True, but sending the args to all PP is the current behaviour. So we must maintain an option to still do it. We could maybe discourage the user from using default by either:
> 
>     1. Providing a warning when default is used
> 
>     2. Leaving the default option undocumented
> 
>     3. Removing the default prefix. Note that the default args can still be set by using the older `--postprocessor-args ARGS` syntax
> 
> 
> imo, a simple warning is sufficient. But I will go with you guys' decision

passing arguments to all post-processor is still allowed by not using any prefix and it will be kept only for backward compatibility(ex: users already having the option in their config file).
so i think option 3 would be best way, for future use the user will be encouraged to only use the specific prefixes and there won't be a `default` prefix to be used only the old way without specifiying a prefix(a warning here that this usage is deprecated and might be removed in the future can be usefull).

pukkandan commented:
@remitamine I will implement your suggestions tonight and let you know

pukkandan commented:
I added the ability to use general program prefix like `ffmpeg`. 

Additionally, you can also give the arguments much more specifically by using `<PP>+<exe>:<args>`. For example, you can give `FFmpegExtractAudio+FFmpeg:` and `FFmpegExtractAudio+FFprobe:` seperately. Suggestions are welcome for a different syntax to do this.

I have not yet updated the documentation to reflect these changes. Honestly, I am not entirely sure how to explain the feature in a simple way. Please give me suggestions if you have any

Also, I noticed that EmbedThumbnailPP (when using AtomicParsley), MetadataFromTitlePP and XAttrMetadataPP currently ignore `--postprocessor-args`.
* I have added support for `AtomicParsley`
* `MetadataFromTitle` does not invoke any external application, so the option does not make sense for it
* I dont understand `XAttrMetadata` enough to add support.

remitamine commented:
as it now allowed to use the `--postprocessor-args` option multiple times, it would be better to see how to handle:

- the same option(`--postprocessor-args`) repeated multiple times with
  - the same prefix but with diffirent arguments(ex: `EmbedThumbnail` used multiple times)
  - with mutiple prefix levels(ex: prefix-less arguments with specfic arguments in the same command)
- multiple options passed to the post-processor programs(same option passed multiple times to ffmpeg or AtomicParsly with different values).

pukkandan commented:
> the same prefix but with diffirent arguments(ex: EmbedThumbnail used multiple times)

The same way all switches are handled. The latter arguments override the former

> multiple options passed to the post-processor programs(same option passed multiple times to ffmpeg or AtomicParsly with different values).

If you pass it with a single switch, it is passed as-is. If you pass it seperately, the latter takes priority. For example, 
`--post-processor-args "ffmpeg:-map 0 -map 1"` = `'-map 0 -map 1`
`--post-processor-args "ffmpeg:-map 0" --post-processor-args  "ffmpeg:-map 1"` = `'-map 1`
This is consistent with how other options work.

> with mutiple prefix levels(ex: prefix-less arguments with specfic arguments in the same command)

I don't understand what you mean. Could you provide example?

remitamine commented:
> I don't understand what you mean. Could you provide example?

`--post-processor-args "ffmpeg:-map 0" --post-processor-args "-map 1"`

pukkandan commented:
>--post-processor-args "ffmpeg:-map 0" --post-processor-args "-map 1"

hm.. I looked through the code again and this case is not properly handled. I will make a commit to fix it.

Here's how I think this should work: (`--ppa` = `--postprocessor-args` - I dont wanna type it every time)
Eg: `--ppa "EmbedThumbnail: et_args" --ppa "ffmpeg: ff_args" --ppa "EmbedThumbnail+ffprobe: probe_args --ppa "def_args"`

should result in:
* ffprobe in EmbedThumbnail gets `probe_args`
* ffmpeg in EmbedThumbnail gets `et_args + ff_args`
* Atomicparsley in Embedthumbnail gets `et_args`
* All other ffmpeg gets `ff_args`
* Any other PP gets `def_args`

Let me know if this a reasonable approach or changes need to be made

remitamine commented:
> Any other PP gets def_args

i think to preserve the current behaviour, the arguments would be passed to all PPs including the ones that are using specific prefixes.

pukkandan commented:
> i think to preserve the current behaviour, the arguments would be passed to all PPs including the ones that are using specific prefixes.

I don't think it should work like that. It shouldn't be a issue that the new syntax, when used, is overriding the old one. Also, this allows passing args easily to all PP except a specific one. Eg: `--ppa "def_args" --ppa "ffprobe:"`

remitamine commented:
> > i think to preserve the current behaviour, the arguments would be passed to all PPs including the ones that are using specific prefixes.
> 
> I don't think it should work like that. It shouldn't be a issue that the new syntax, when used, is overriding the old one. Also, this allows passing args easily to all PP except a specific one. Eg: `--ppa "def_args" --ppa "ffprobe:"`

this inconsitent with the behaviour that you've described in you're example:

> ffmpeg in EmbedThumbnail gets et_args + ff_args

where the ffmpeg command will combine the program specific args and the PP specific args, while in the example i described the options are not combined.

pukkandan commented:
> --post-processor-args "ffmpeg:-map 0" --post-processor-args "-map 1"
> while in the example i described the options are not combined.

There is no PP specific args given in this

> this inconsitent with the behaviour that you've described in you're example:

No, it is exactly how I specified it

> --ppa "EmbedThumbnail: et_args" --ppa "ffmpeg: ff_args" --ppa "EmbedThumbnail+ffprobe: probe_args" --ppa "def_args"
ffprobe in EmbedThumbnail gets probe_args
ffmpeg in EmbedThumbnail gets et_args + ff_args
Atomicparsley in Embedthumbnail gets et_args
All other ffmpeg gets ff_args
Any other PP gets def_args

If `PP+exe` args is specifically given, only it used.  
Otherwise, PP and exe args are combined.
If none of the `PP`, `exe` or `PP+exe` are given, the default is used

Please explain what you find inconsistent with it

remitamine commented:
as i think of the prefixes, i think of them as classes from the most specfic to the more generic:
- `PP+exe`
- `PP`
- `exe`
- ``(prefixless)
but i think you're considiring `PP` and `exe` in the same level.
in the way you're classifying the behviour would be consitent(but not backward compatible), but not in the way that i'm classifying the prefixes.
so it would come to:
- what classification is more correct.
- what behaviour would be better(overiding the generic args with the more specific or combine them) or to allow both of them.

pukkandan commented:
> but i think you're considering PP and exe in the same level.

Yes, there is no reasonable way to say which of PP/exe is more specific. `AtomicParsley` is more specific than `EmbedThumbnail` which in turn (arguably) has more specificity than `ffmpeg`. So considering them to be at the same level and combining them makes sense imo.

First I also thought of simply combining all the args. This is also the easiest to code. But it would make it very difficult for the user to override previous args.
> Also, this allows passing args easily to all PP except a specific one. Eg: `--ppa "def_args" --ppa "ffprobe:"`

If you still think that is the way to go, I can modify the code to be so.

remitamine commented:
> Yes, there is no reasonable way to say which of PP/exe is more specific. AtomicParsley is more specific than EmbedThumbnail which in turn (arguably) has more specificity than ffmpeg. So considering them to be at the same level and combining them makes sense imo.

a program is not bound to only a specific PP, there nothing that prevents `AtomicParsley` from been used in another PP, so that's why i think the program prefix is more generic than `PP` prefix, while in the case of a PP using multiple programs, it is more specific then `PP` prefix and it's covered by `PP+exe` prefix.

pukkandan commented:
> a program is not bound to only a specific PP

I could argue the same in reverse. But that is beside the point. Honestly, what we are discussing is simply convention and I don't have any strong preference either way. As one of the core maintainers of this project, it is up to you to decide how it should behave. Any of the behaviors we discussed are easy to implement and not too different in usage. We simply need to decide on a convention.

Maybe discuss with @dstftw and finalize how the behaviour should be, I and I will make it so.

remitamine commented:
> I could argue the same in reverse. But that is beside the point. Honestly, what we are discussing is simply convention and I don't have any strong preference either way. As one of the core maintainers of this project, it is up to you to decide how it should behave. Any of the behaviors we discussed are easy to implement and not too different in usage. We simply need to decide on a convention.

as i'm not the one who started the review process, i'm not even trying to make a full review here(otherwise i would'nt use just comments and would've used the GitHub requested changes process), again, what i have provided untill now are just suggestions, there where some points that i wanted to be taken into considiration.
i think it's better to think about good design(by discussing) that wouldn't just work now, but try to cover future uses as well, generally when things are rushed anddo whatever that would work just now, it would bring more problems in the future.
anyway, i think the points that i wanted to raise can be found in my previous comments.
